src/mlpack/core/cv/cv_base.hpp

```diff
@@ -0,0 +1,329 @@
+   * @param ys Labels for each point from the dataset.
+   * @param numClasses Number of classes in the dataset.
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   */
+  template<typename MT, typename PT>
+  CVBase(const MT& xs,
+         const data::DatasetInfo& datasetInfo,
+         const PT& ys,
+         const size_t numClasses);
+
+  /*
+   * This constructor can be used for regression and binary classification
+   * algorithms that support weighted learning.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param ys Predictions (labels for classification algorithms and responses
+   *     for regression algorithms) for each point from the dataset.
+   * @param weights Observation weights (for boosting).
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   * @tparam WT A type that can be converted to WeightsType.
+   */
+  template<typename MT, typename PT, typename WT>
+  CVBase(const MT& xs,
+         const PT& ys,
+         const WT& weights);
+
+  /*
+   * This constructor can be used for multiclass classification algorithms that
+   * support weighted learning.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param ys Labels for each point from the dataset.
+   * @param numClasses Number of classes in the dataset.
+   * @param weights Observation weights (for boosting).
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   * @tparam WT A type that can be converted to WeightsType.
+   */
+  template<typename MT, typename PT, typename WT>
+  CVBase(const MT& xs,
+         const PT& ys,
+         const size_t numClasses,
+         const WT& weights);
+
+  /*
+   * This constructor can be used for multiclass classification algorithms that
+   * can take a data::DatasetInfo parameter and support weighted learning.
+   *
+   * @param xs Dataset to cross-validate on.
+   * @param datasetInfo Type information for each dimension of the dataset.
+   * @param ys Labels for each point from the dataset.
+   * @param numClasses Number of classes in the dataset.
+   * @param weights Observation weights (for boosting).
+   *
+   * @tparam MT A type that can be converted to MatType.
+   * @tparam PT A type that can be converted to PredictionsType.
+   * @tparam WT A type that can be converted to WeightsType.
+   */
+  template<typename MT, typename PT, typename WT>
+  CVBase(const MT& xs,
+         const data::DatasetInfo& datasetInfo,
+         const PT& ys,
+         const size_t numClasses,
+         const WT& weights);
+
+ protected:
+  using MIE =
+      MetaInfoExtractor<MLAlgorithm, MatType, PredictionsType, WeightsType>;
+
+  static_assert(MIE::IsSupported,
+      "MLAlgorithm should be supported by MetaInfoExtractor");
```

Comment:
Yeah---I think `should` leaves some ambiguity about whether the message is an internal warning, a warning to the user, or an actual error.  I think it is better to be clear that the error is caused by the user explicitly using the class wrong.